README.md
# WTUIColorExtensions

[![CI Status](http://img.shields.io/travis/wltrup/iOS-Swift-WTUIColorExtensions.svg?style=flat)](https://travis-ci.org/wltrup/iOS-Swift-WTUIColorExtensions)
[![Version](https://img.shields.io/cocoapods/v/WTUIColorExtensions.svg?style=flat)](http://cocoapods.org/pods/WTUIColorExtensions)
[![License](https://img.shields.io/cocoapods/l/WTUIColorExtensions.svg?style=flat)](http://cocoapods.org/pods/WTUIColorExtensions)
[![Platform](https://img.shields.io/cocoapods/p/WTUIColorExtensions.svg?style=flat)](http://cocoapods.org/pods/WTUIColorExtensions)

## What

**WTUIColorExtensions** adds to `UIColor` the ability to generate pseudo-random color instances
and methods to extract color components in a more direct way than using `inout` arguments and
the default `get...` methods. Here's the full set of API additions:

```swift
    public static var randomRGB: UIColor { get }
    public static var randomRGBA: UIColor { get }

    public static var randomHSB: UIColor { get }
    public static var randomHSBA: UIColor { get }

    public static var randomWhite: UIColor { get }
    public static var randomWhiteAlpha: UIColor { get }

    public enum ColorError : Error {
        case invalidColorSpace
    }

    public struct RGBAComponents {
        public let red: CGFloat
        public let green: CGFloat
        public let blue: CGFloat
        public let alpha: CGFloat
    }
    public func rgbaComponents() throws -> RGBAComponents

    public struct HSBAComponents {
        public let hue: CGFloat
        public let saturation: CGFloat
        public let brightness: CGFloat
        public let alpha: CGFloat
    }
    public func hsbaComponents() throws -> HSBAComponents

    public struct WhiteAlphaComponents {
        public let white: CGFloat
        public let alpha: CGFloat
    }
    public func whiteAlphaComponents() throws -> WhiteAlphaComponents

    // NEW in v. 1.1.0
    public func hexValue() throws -> String
    public func hexValueWithAlpha() throws -> String

    // NEW in v. 1.1.0
    public static func quadraticLuma(red: CGFloat, green: CGFloat, blue: CGFloat) throws -> CGFloat
    public func quadraticLuma() throws -> CGFloat

    // NEW in v. 1.1.0
    public func contrastingColor(continuous: Bool, threshold: CGFloat, sameAlpha: Bool) throws -> UIColor
}
```

## Why

Because generating random colors is a common enough need and because having to use `inout`
arguments to extract color components is annoying.

## Tests

**WTUIColorExtensions** adds several extensions, all tested by 26 tests with nearly 95% coverage.
The missing 5% test-coverage is due to the fact that `rgbaComponents()` and `hsbaComponents()`
don't actually `throw` any invalid color space errors on iOS, although in theory they could.
In order to future-proof this library, I've assumed that those methods might fail but I can't
currently test that code path.

## Changelog

Changes to **WTUIColorExtensions** are listed
[here](https://github.com/wltrup/iOS-Swift-WTUIColorExtensions/blob/master/CHANGELOG.md).

## Installation

**WTUIColorExtensions** is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "WTUIColorExtensions"
```

## Author

Wagner Truppel, trupwl@gmail.com

## License

**WTUIColorExtensions** is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# WTUIColorExtensions

## 1.1.1

- No functional changes. Just fixes to the README file.

## 1.1.0

- `whiteAlphaComponents()` no longer throws when the RBG components aren't equal to each other. 
- Added `hexValue()` and `hexValueWithAlpha()` for finding the hex representation of a color.
- Added `quadraticLuma()` and `quadraticLuma(red: CGFloat, green: CGFloat, blue: CGFloat)` for
    obtaining the value of a color's luminance, using a quadratic approximation'.
- Added `contrastingColor(continuous: Bool, threshold: CGFloat, sameAlpha: Bool)` to find the
    most contrasting color to a given color.

## 1.0.2

- Aligns versioning with other of my libraries.

## 1.0.1

- Adds this changelog file but no other functional changes.

## 1.0.0

- Initial release.


Example/Tests/Tests.swift
import UIKit
import XCTest
import WTUIColorExtensions

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/Tests/WTUIColorExtensionsTests.swift
//
//  WTUIColorExtensionsTests.swift
//  WTUIColorExtensions
//
//  Created by Wagner Truppel on 2016.12.03.
//
//  Copyright (c) 2016 Wagner Truppel. All rights reserved.
//

import XCTest
import Foundation
import WTUIColorExtensions

class WTUIColorExtensionsTests: WTUIColorExtensionsTestsBase
{
    func test_randomRGBGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomRGB.rgbaComponents()
                return (components.red   >= 0 && components.red   <= 1 &&
                    components.green >= 0 && components.green <= 1 &&
                    components.blue  >= 0 && components.blue  <= 1 &&
                    components.alpha == 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomRGB()
    {
        N = 10_000
        tolerance = 1e-1

        let colors = (1...N)
            .map { _ in try! UIColor.randomRGB.rgbaComponents() }

        let reds = colors.map { $0.red }
        testRandomness(reds)

        let greens = colors.map { $0.green }
        testRandomness(greens)

        let blues = colors.map { $0.blue }
        testRandomness(blues)
    }

    // MARK: -

    func test_randomRGBAGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomRGBA.rgbaComponents()
                return (components.red   >= 0 && components.red   <= 1 &&
                    components.green >= 0 && components.green <= 1 &&
                    components.blue  >= 0 && components.blue  <= 1 &&
                    components.alpha >= 0 && components.alpha <= 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomRGBA()
    {
        N = 10_000
        tolerance = 1e-1

        let colors = (1...N)
            .map { _ in try! UIColor.randomRGBA.rgbaComponents() }

        let reds = colors.map { $0.red }
        testRandomness(reds)

        let greens = colors.map { $0.green }
        testRandomness(greens)

        let blues = colors.map { $0.blue }
        testRandomness(blues)

        let alphas = colors.map { $0.alpha }
        testRandomness(alphas)
    }

    // MARK: -

    func test_randomHSBGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomHSB.hsbaComponents()
                return (components.hue         >= 0 && components.hue         <= 1 &&
                    components.saturation  >= 0 && components.saturation  <= 1 &&
                    components.brightness  >= 0 && components.brightness  <= 1 &&
                    components.alpha       == 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomHSB()
    {
        N = 10_000
        tolerance = 1e-1

        let colors = (1...N)
            .map { _ in try! UIColor.randomHSB.hsbaComponents() }

        let hues = colors.map { $0.hue }
        testRandomness(hues)

        let sats = colors.map { $0.saturation }
        testRandomness(sats)

        let brights = colors.map { $0.brightness }
        testRandomness(brights)
    }

    // MARK: -

    func test_randomHSBAGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomHSBA.hsbaComponents()
                return (components.hue         >= 0 && components.hue         <= 1 &&
                    components.saturation  >= 0 && components.saturation  <= 1 &&
                    components.brightness  >= 0 && components.brightness  <= 1 &&
                    components.alpha       >= 0 && components.alpha       <= 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomHSBA()
    {
        N = 10_000
        tolerance = 1e-1

        let colors = (1...N)
            .map { _ in try! UIColor.randomHSBA.hsbaComponents() }

        let hues = colors.map { $0.hue }
        testRandomness(hues)

        let sats = colors.map { $0.saturation }
        testRandomness(sats)

        let brights = colors.map { $0.brightness }
        testRandomness(brights)

        let alphas = colors.map { $0.alpha }
        testRandomness(alphas)
    }

    // MARK: -

    func test_randomWhiteGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomWhite.whiteAlphaComponents()
                return (components.white >= 0 && components.white <= 1 &&
                    components.alpha == 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomWhite()
    {
        N = 10_000
        tolerance = 1e-1

        let whites = (1...N)
            .map { _ in try! UIColor.randomWhite.whiteAlphaComponents().white }

        testRandomness(whites)
    }

    // MARK: -

    func test_randomWhiteAlphaGeneratesColorsWithComponentValuesInTheExpectedRange()
    {
        let success = (1...N)
            .map { _ in
                let components = try! UIColor.randomWhiteAlpha.whiteAlphaComponents()
                return (components.white >= 0 && components.white <= 1 &&
                    components.alpha >= 0 && components.alpha <= 1)
            }
            .reduce(true) { $0 && $1 }

        XCTAssertTrue(success)
    }

    func test_randomWhiteAlpha()
    {
        N = 10_000
        tolerance = 1e-1

        let colors = (1...N)
            .map { _ in try! UIColor.randomWhiteAlpha.whiteAlphaComponents() }

        let whites = colors.map { $0.white }
        testRandomness(whites)

        let alphas = colors.map { $0.alpha }
        testRandomness(alphas)
    }

    // MARK: -

    func test_rgbaComponentsDoesNotActuallyThrow1()
    {
        let color = UIColor(hue: random01(),
                            saturation: random01(),
                            brightness: random01(),
                            alpha: random01())

        do { _ = try color.rgbaComponents() }
        catch { XCTFail() }
    }

    func test_rgbaComponentsDoesNotActuallyThrow2()
    {
        let color = UIColor(white: random01(), alpha: random01())

        do { _ = try color.rgbaComponents() }
        catch { XCTFail() }
    }

    func test_rgbaComponents()
    {
        (1...N).forEach { _ in
            let r = random01(); let g = random01()
            let b = random01(); let a = random01()

            let color = UIColor(red: r, green: g, blue: b, alpha: a)
            let comps = try! color.rgbaComponents()

            expectedValue = r
            resultedValue = comps.red
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = g
            resultedValue = comps.green
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = b
            resultedValue = comps.blue
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = a
            resultedValue = comps.alpha
            assertAbsoluteDifferenceWithinTolerance()
        }
    }

    // MARK: -

    func test_hsbaComponentsDoesNotActuallyThrow1()
    {
        let color = UIColor(red: random01(),
                            green: random01(),
                            blue: random01(),
                            alpha: random01())

        do { _ = try color.hsbaComponents() }
        catch { XCTFail() }
    }

    func test_hsbaComponentsDoesNotActuallyThrow2()
    {
        let color = UIColor(white: random01(), alpha: random01())

        do { _ = try color.hsbaComponents() }
        catch { XCTFail() }
    }

    func test_hsbaComponents()
    {
        (1...N).forEach { _ in
            let h = random01(); let s = random01()
            let b = random01(); let a = random01()

            let color = UIColor(hue: h, saturation: s, brightness: b, alpha: a)
            let comps = try! color.hsbaComponents()

            expectedValue = h
            resultedValue = comps.hue
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = s
            resultedValue = comps.saturation
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = b
            resultedValue = comps.brightness
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = a
            resultedValue = comps.alpha
            assertAbsoluteDifferenceWithinTolerance()
        }
    }

    // MARK: -

    func test_whiteAlphaComponents()
    {
        (1...N).forEach { _ in
            let w = random01(); let a = random01()

            let color = UIColor(white: w, alpha: a)
            let comps = try! color.whiteAlphaComponents()

            expectedValue = w
            resultedValue = comps.white
            assertAbsoluteDifferenceWithinTolerance()

            expectedValue = a
            resultedValue = comps.alpha
            assertAbsoluteDifferenceWithinTolerance()
        }
    }

    // MARK: -

    fileprivate func hexDigit(_ n: Int) -> String
    {
        if n < 10 { return "\(n)" }
        if n == 10 { return "A" }
        if n == 11 { return "B" }
        if n == 12 { return "C" }
        if n == 13 { return "D" }
        if n == 14 { return "E" }
        if n == 15 { return "F" }
        return ""
    }

    fileprivate func hexInteger(_ n: Int) -> String
    {
        guard n > 0 else { return "00" }

        var a: [String] = []
        var m = n
        while m > 0 {
            a = [hexDigit(m % 16)] + a
            m /= 16
        }

        let p = (n > 15 ? "" : "0")
        return p + a.reduce("", +)
    }

    func test_hexValueWithAlpha()
    {
        (1...N).forEach { _ in
            let r = random01(); let rn = Int(255 * r); let rhex = hexInteger(rn)
            let g = random01(); let gn = Int(255 * g); let ghex = hexInteger(gn)
            let b = random01(); let bn = Int(255 * b); let bhex = hexInteger(bn)
            let a = random01(); let an = Int(255 * a); let ahex = hexInteger(an)

            let color = UIColor(red: r, green: g, blue: b, alpha: a)
            let expectedValue = "\(rhex)\(ghex)\(bhex)\(ahex)"
            let resultedValue = try! color.hexValueWithAlpha()

            XCTAssertEqual(resultedValue, expectedValue)
        }
    }

    func test_hexValue()
    {
        (1...N).forEach { _ in
            let r = random01(); let rn = Int(255 * r); let rhex = hexInteger(rn)
            let g = random01(); let gn = Int(255 * g); let ghex = hexInteger(gn)
            let b = random01(); let bn = Int(255 * b); let bhex = hexInteger(bn)
            let a = random01()

            let color = UIColor(red: r, green: g, blue: b, alpha: a)
            let expectedValue = "\(rhex)\(ghex)\(bhex)"
            let resultedValue = try! color.hexValue()

            XCTAssertEqual(resultedValue, expectedValue)
        }
    }

    // MARK: -

    func test_quadraticLuma_static()
    {
        (1...N).forEach { _ in
            let r = random01(); let g = random01(); let b = random01()
            let expectedValue = (0.2126 * r * r + 0.7152 * g * g + 0.0722 * b * b)
            let resultedValue = try! UIColor.quadraticLuma(red: r, green: g, blue: b)

            XCTAssertEqualWithAccuracy(resultedValue, expectedValue, accuracy: 1e-8)
        }
    }

    func test_quadraticLuma()
    {
        (1...N).forEach { _ in
            let r = random01(); let g = random01()
            let b = random01(); let a = random01()
            let color = UIColor(red: r, green: g, blue: b, alpha: a)
            let expectedValue = (0.2126 * r * r + 0.7152 * g * g + 0.0722 * b * b)
            let resultedValue = try! color.quadraticLuma()

            XCTAssertEqualWithAccuracy(resultedValue, expectedValue, accuracy: 1e-8)
        }
    }

    // MARK: -

    func test_contrastingColor_1()
    {
        for continuous in [false, true] {
            for sameAlpha in [false, true] {

                let original = UIColor.black
                let expected = UIColor.white
                let resulted = try! original.contrastingColor(continuous: continuous,
                                                              sameAlpha: sameAlpha)
                XCTAssertEqual(resulted, expected)

            }
        }
    }

    func test_contrastingColor_2()
    {
        for continuous in [false, true] {
            for sameAlpha in [false, true] {

                let original = UIColor.white
                let expected = UIColor.black
                let resulted = try! original.contrastingColor(continuous: continuous,
                                                              sameAlpha: sameAlpha)
                XCTAssertEqual(resulted, expected)

            }
        }
    }

    func test_contrastingColor_3()
    {
        (1...N).forEach { _ in
            let r = random01(); let g = random01()
            let b = random01(); let a = random01()
            let original = UIColor(red: r, green: g, blue: b, alpha: a)
            
            let luma = try! UIColor.quadraticLuma(red: r, green: g, blue: b)
            
            let expected: UIColor
            if luma > 0.2 {
                expected = UIColor.black.withAlphaComponent(a)
            }
            else {
                expected = UIColor.white.withAlphaComponent(a)
            }
            let resulted = try! original.contrastingColor(continuous: false,
                                                          sameAlpha: true)
            
            XCTAssertEqual(resulted, expected)
        }
    }
    
}



WTUIColorExtensions/Classes/WTUIColorExtensions.swift
/*
 WTUIColorExtensions.swift
 WTUIColorExtensions

 Created by Wagner Truppel on 2016.12.03.

 The MIT License (MIT)

 Copyright (c) 2016 Wagner Truppel.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 When crediting me (Wagner Truppel) for this work, please use one
 of the following two suggested formats:

 Uses "WTUIColorExtensions" by Wagner Truppel
 https://github.com/wltrup

 or

 WTUIColorExtensions by Wagner Truppel
 https://github.com/wltrup
 */

import Foundation
import CoreGraphics
import UIKit


public extension UIColor
{
    /// Returns a new `UIColor` instance in the **RGB color space**, with
    /// pseudo-random values for its color coordinates (red, green, blue),
    /// but with its opacity (alpha) set to 1.
    ///
    /// - SeeAlso: `UIColor.randomRGBA`.
    ///
    /// - SeeAlso: `UIColor.randomHSB`.
    /// - SeeAlso: `UIColor.randomHSBA`.
    ///
    /// - SeeAlso: `UIColor.randomWhite`.
    /// - SeeAlso: `UIColor.randomWhiteAlpha`.
    public static var randomRGB: UIColor
    {
        let r = UIColor.random01
        let g = UIColor.random01
        let b = UIColor.random01
        return UIColor(red: r, green: g, blue: b, alpha: 1)
    }

    /// Returns a new `UIColor` instance in the **RGB color space**, with
    /// pseudo-random values for all its color coordinates (red, green, blue, alpha).
    ///
    /// - SeeAlso: `UIColor.randomRGB`.
    ///
    /// - SeeAlso: `UIColor.randomHSB`.
    /// - SeeAlso: `UIColor.randomHSBA`.
    ///
    /// - SeeAlso: `UIColor.randomWhite`.
    /// - SeeAlso: `UIColor.randomWhiteAlpha`.
    public static var randomRGBA: UIColor
    {
        let r = UIColor.random01
        let g = UIColor.random01
        let b = UIColor.random01
        let a = UIColor.random01
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }

    // MARK: -

    /// Returns a new `UIColor` instance in the **HSB color space**,\* with
    /// pseudo-random values for its color coordinates (hue, saturation, brightness),
    /// but with its opacity (alpha) set to 1.
    ///
    /// \* NOTE: Although this computed property creates a `UIColor` instance
    ///          using HSB values, that instance is internally stored as an
    ///          RGB color so it's not actually in the HSB color space.
    ///
    /// - SeeAlso: `UIColor.randomRGB`.
    /// - SeeAlso: `UIColor.randomRGBA`.
    ///
    /// - SeeAlso: `UIColor.randomHSBA`.
    ///
    /// - SeeAlso: `UIColor.randomWhite`.
    /// - SeeAlso: `UIColor.randomWhiteAlpha`.
    public static var randomHSB: UIColor
    {
        let h = UIColor.random01
        let s = UIColor.random01
        let b = UIColor.random01
        return UIColor(hue: h, saturation: s, brightness: b, alpha: 1)
    }

    /// Returns a new `UIColor` instance in the **HSB color space**,\* with
    /// pseudo-random values for all its color coordinates
    /// (hue, saturation, brightness, alpha).
    ///
    /// \* NOTE: Although this computed property creates a `UIColor` instance
    ///          using HSB values, that instance is internally stored as an
    ///          RGB color so it's not actually in the HSB color space.
    ///
    /// - SeeAlso: `UIColor.randomRGB`.
    /// - SeeAlso: `UIColor.randomRGBA`.
    ///
    /// - SeeAlso: `UIColor.randomHSB`.
    ///
    /// - SeeAlso: `UIColor.randomWhite`.
    /// - SeeAlso: `UIColor.randomWhiteAlpha`.
    public static var randomHSBA: UIColor
    {
        let h = UIColor.random01
        let s = UIColor.random01
        let b = UIColor.random01
        let a = UIColor.random01
        return UIColor(hue: h, saturation: s, brightness: b, alpha: a)
    }

    // MARK: -

    /// Returns a new `UIColor` instance in the **grayscale**\* color space,
    /// with a pseudo-random value for its white coordinate but with its
    /// opacity (alpha) set to 1.
    ///
    /// \* NOTE: Although this computed property creates a `UIColor` instance
    ///          using a grayscale value, that instance is internally stored
    ///          as an RGB color so it's not actually in the grayscale color
    ///          space.
    ///
    /// - SeeAlso: `UIColor.randomRGB`.
    /// - SeeAlso: `UIColor.randomRGBA`.
    ///
    /// - SeeAlso: `UIColor.randomHSB`.
    /// - SeeAlso: `UIColor.randomHSBA`.
    ///
    /// - SeeAlso: `UIColor.randomWhiteAlpha`.
    public static var randomWhite: UIColor
    {
        let w = UIColor.random01
        return UIColor(white: w, alpha: 1)
    }

    /// Returns a new `UIColor` instance in the **grayscale color space**,\*
    /// with a pseudo-random value for both its white coordinate and its
    /// opacity coordinate (alpha).
    ///
    /// \* NOTE: Although this computed property creates a `UIColor` instance
    ///          using a grayscale value, that instance is internally stored
    ///          as an RGB color so it's not actually in the grayscale color
    ///          space.
    ///
    /// - SeeAlso: `UIColor.randomRGB`.
    /// - SeeAlso: `UIColor.randomRGBA`.
    ///
    /// - SeeAlso: `UIColor.randomHSB`.
    /// - SeeAlso: `UIColor.randomHSBA`.
    ///
    /// - SeeAlso: `UIColor.randomWhite`.
    public static var randomWhiteAlpha: UIColor
    {
        let w = UIColor.random01
        let a = UIColor.random01
        return UIColor(white: w, alpha: a)
    }

    // MARK: -

    /// An enumeration describing the possible errors that can be thrown when using
    /// functions from the extended `UIColor` API provided by **WTUIColorExtensions**.
    ///
    /// - **invalidColorSpace**:
    ///            Signifies that the attempted function cannot be performed with the
    ///            color space provided.
    public enum ColorError: Error
    {
        case invalidColorSpace
    }

    // MARK: -

    /// A struct to store the RGBA components of a `UIColor` instance.
    public struct RGBAComponents
    {
        public let red: CGFloat
        public let green: CGFloat
        public let blue: CGFloat
        public let alpha: CGFloat
    }

    /// Returns the RGBA components of `self`.
    ///
    /// - Returns: the RGBA components of `self`.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public func rgbaComponents() throws -> RGBAComponents
    {
        var r: CGFloat = 0; var g: CGFloat = 0;
        var b: CGFloat = 0; var a: CGFloat = 0
        let success = getRed(&r, green: &g, blue: &b, alpha: &a)
        guard success else { throw ColorError.invalidColorSpace }
        return RGBAComponents(red: r, green: g, blue: b, alpha: a)
    }

    // MARK: -

    /// A struct to store the HSBA components of a `UIColor` instance.
    public struct HSBAComponents
    {
        public let hue: CGFloat
        public let saturation: CGFloat
        public let brightness: CGFloat
        public let alpha: CGFloat
    }

    /// Returns the HSBA components of `self`.
    ///
    /// - Returns: the HSBA components of `self`.
    ///
    /// - Throws: `Error.invalidColorSpace` if somehow
    ///           the HSBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public func hsbaComponents() throws -> HSBAComponents
    {
        var h: CGFloat = 0; var s: CGFloat = 0;
        var b: CGFloat = 0; var a: CGFloat = 0
        let success = getHue(&h, saturation: &s, brightness: &b, alpha: &a)
        guard success else { throw ColorError.invalidColorSpace }
        return HSBAComponents(hue: h, saturation: s, brightness: b, alpha: a)
    }

    // MARK: -

    /// A struct to store the white and alpha components of a `UIColor` instance
    /// in the grayscale color space.
    public struct WhiteAlphaComponents
    {
        public let white: CGFloat
        public let alpha: CGFloat
    }

    /// Returns the white and alpha components of `self`.
    ///
    /// - Returns: the white and alpha components of `self`.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the white and alpha components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public func whiteAlphaComponents() throws -> WhiteAlphaComponents
    {
        var w: CGFloat = 0; var a: CGFloat = 0
        let success = getWhite(&w, alpha: &a)
        guard success else { throw ColorError.invalidColorSpace }

        return WhiteAlphaComponents(white: w, alpha: a)
    }

    // MARK: -

    /// Returns the hexadecimal representation of an RGBA color, as a string,
    /// **excluding** its `alpha` component.
    ///
    /// - Returns: the hexadecimal representation of `self`, as a string,
    ///            **excluding** its `alpha` component.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    ///
    /// - SeeAlso: hexValueWithAlpha()
    public func hexValue() throws -> String
    {
        let rgba = try rgbaComponents()
        return [rgba.red, rgba.green, rgba.blue]
            .map { Int(255 * $0) }
            .map { String(format:"%02X", $0) }
            .reduce("", +)
    }

    /// Returns the hexadecimal representation of an RGBA color, as a string,
    /// **including** its `alpha` component.
    ///
    /// - Returns: the hexadecimal representation of `self`, as a string,
    ///            **including** its `alpha` component.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    ///
    /// - SeeAlso: hexValue()
    public func hexValueWithAlpha() throws -> String
    {
        let rgba = try rgbaComponents()
        return [rgba.red, rgba.green, rgba.blue, rgba.alpha]
            .map { Int(255 * $0) }
            .map { String(format:"%02X", $0) }
            .reduce("", +)
    }

    // MARK: -

    /// Returns an approximate value for the **luminance** of an RGB color,
    /// based on the quadratic expression:
    ///
    /// `L = 0.2126 * red * red + 0.7152 * green * green + 0.0722 * blue * blue`
    ///
    /// where `(red, green, blue)`, and `L` as well, are all in the range `[0, 1]`.
    ///
    /// See:
    /// http://gamedev.stackexchange.com/questions/38536/given-a-rgb-color-x-how-to-find-the-most-contrasting-color-y
    ///
    /// - Parameters:
    ///   - red:   the red   component of `self`, in the range [0,1].
    ///   - green: the green component of `self`, in the range [0,1].
    ///   - blue:  the blue  component of `self`, in the range [0,1].
    ///
    /// - Returns: an approximate value for the **luminance** of a color with the
    ///            given RGB values.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public static func quadraticLuma(red: CGFloat,
                                     green: CGFloat,
                                     blue: CGFloat) throws -> CGFloat
    {
        let rsq = red * red
        let gsq = green * green
        let bsq = blue * blue
        return (0.2126 * rsq + 0.7152 * gsq + 0.0722 * bsq)
    }

    /// Returns an approximate value for the **luminance** of an RGB color,
    /// based on the quadratic expression:
    ///
    /// `L = 0.2126 * red * red + 0.7152 * green * green + 0.0722 * blue * blue`
    ///
    /// where `(red, green, blue)`, and `L` as well, are all in the range `[0, 1]`.
    ///
    /// See:
    /// http://gamedev.stackexchange.com/questions/38536/given-a-rgb-color-x-how-to-find-the-most-contrasting-color-y
    ///
    /// - Returns: an approximate value for the **luminance** of `self`.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public func quadraticLuma() throws -> CGFloat
    {
        let rgba = try rgbaComponents()
        return try UIColor.quadraticLuma(red: rgba.red,
                                         green: rgba.green,
                                         blue: rgba.blue)
    }

    // MARK: -

    /// Returns a color that contrasts well with `self`.
    ///
    /// See:
    /// http://gamedev.stackexchange.com/questions/38536/given-a-rgb-color-x-how-to-find-the-most-contrasting-color-y
    ///
    /// - Parameters:
    ///
    ///   - continuous: when `true`, the resulting color is some shade of gray ranging
    ///                 from white to black; when `false`, the resulting color is either
    ///                 white or black, depending on a luminance threshold. The default
    ///                 behaviour is to result in a continuous shade of gray.
    ///
    ///   - threshold: ignored when `continuous` is set to `true`, otherwise used to
    ///                decide whether the resulting color is white or black. The result
    ///                is black when `self` has a luminance larger than this threshold,
    ///                else the result if white. The default threshold value is 0.2.
    ///
    ///   - sameAlpha: when `true`, the resulting color has the same `alpha` component
    ///                as `self`, otherwise it has `alpha` equal to 1. The default is
    ///                `false` so the resulting color has `alpha` equal to 1.
    ///
    /// - Returns: a color that contrasts well with `self`.
    ///
    /// - Throws: `ColorError.invalidColorSpace` if somehow
    ///           the RGBA components can't be extracted from `self` (this
    ///           should never happen on iOS but...)
    public func contrastingColor(continuous: Bool = true,
                                 threshold: CGFloat = 0.2,
                                 sameAlpha: Bool = false) throws -> UIColor
    {
        let rgba = try rgbaComponents()
        let luma = try UIColor.quadraticLuma(red: rgba.red,
                                             green: rgba.green,
                                             blue: rgba.blue)
        let a: CGFloat = (sameAlpha ? rgba.alpha : 1)
        if continuous
        { return UIColor(white: 1 - luma, alpha: a) }
        else
        {
            return (luma > threshold ?
                UIColor.black.withAlphaComponent(a) :
                UIColor.white.withAlphaComponent(a))
        }
    }

    // MARK: - Private API

    /// Returns a uniformly-distributed pseudo-random value in the
    /// **closed** interval [0, 1].
    fileprivate static var random01: CGFloat
    { return CGFloat(arc4random_uniform(UInt32.max)) / CGFloat(UInt32.max - 1) }
}

